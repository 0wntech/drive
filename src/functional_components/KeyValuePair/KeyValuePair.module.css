.key {
    width: 33%;
    font-weight: 600;
    color: #aaa;
    text-align: right;
    padding-right: 50px;
    font-size: 1em;
    height: 40px;
}

.values {
    width: 55%;
    font-size: 1em;
    color: #0c89c6;
}

.value {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.section {
    margin: 3px 0;
    width: 100%;
    padding: 0 40 0 20px;
    background-color: white;
    border-radius: 10px;
    padding: 20px 30px;
    display: flex;
}

.editIcon {
    position: absolute;
    right: 20px;
    width: 20px;
    height: 20px;
    cursor: pointer;
}

.editable {
    -moz-appearance: textfield;
    -webkit-appearance: textfield;
    background-color: white;
    background-color: -moz-field;
    border: 1px solid darkgray;
    box-shadow: 1px 1px 1px 0 lightgray inset;
    box-sizing: border-box;
    -moz-box-sizing: border-box;
    -webkit-box-sizing: border-box;
    padding: 0 3px;
}
