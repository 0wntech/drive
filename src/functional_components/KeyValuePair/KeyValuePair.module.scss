@import '../../styles/constants.scss';

.container {
    box-sizing: border-box;
    border-radius: 20px;
    margin-bottom: 20px;
    width: 50%;
}

.keyValueContainer {
    display: flex;
    flex-direction: column;
}

.valueWrapper {
    width: 100%;
    position: relative;
    margin-bottom: 15px;
}

.valueWrapper:last-child {
    margin-bottom: 0;
}

.deleteIcon {
    position: absolute;
    right: 0px;
    top: 50%;
    transform: translateY(-50%);
}

.containerFooter {
    width: 100%;
    height: 50px;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    border-top: none;
}

.multiValueContainer {
    display: flex;
    flex-direction: column;
}

.keyLabel {
    font-style: normal;
    font-weight: 600;
    font-size: 18px;
    width: calc(150px + 1em);
    margin-bottom: 8px;
}

.addLabel {
    font-style: normal;
    font-weight: 300;
    font-size: 18px;
}

.value {
    width: 100%;
    font-style: normal;
    box-sizing: border-box;
    font-weight: 300;
    font-size: 18px;
    border: initial;
    margin-bottom: 2px;
    outline: none;
    resize: none;
}

.value + .value {
    margin-top: 1em;
}

.value.active::placeholder {
    /* text-decoration: underline;
    text-decoration-color: #6db7db; */
}

.value.active {
    border-radius: initial;
    border-bottom: 2px solid black;
    margin-bottom: initial;
    /* text-decoration: underline; */
}

.value.focus::placeholder {
    color: transparent;
}

.value.active:focus {
    outline: initial;
    border-color: #6db7db;
    caret-color: #6db7db;
    /* text-decoration-color: #6db7db; */
}

.multiple {
    margin-bottom: 10px;
}

.multiple:last-child {
    margin-bottom: 0;
}

.valueContainer {
    width: 100%;
}

@media (max-width: $screen_m) {
    .container {
        width: 80%;
    }

    .keyLabel {
        width: calc(70px + 1em);
        font-size: 16px;
    }

    .value {
        font-size: 16px;
    }
}
