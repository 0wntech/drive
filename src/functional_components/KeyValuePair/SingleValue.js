import React, { useState } from 'react';
import TextareaAutosize from 'react-textarea-autosize';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './KeyValuePair.module.scss';

const SingleValue = ({
    dataKey,
    value,
    editable,
    placeholder,
    setValue,
    className,
    maxInput,
}) => {
    const [fallbackValue, setFallbackValue] = useState(value);

    const onFocusIn = (e) => {
        // make Placeholder visible
        if (editable) {
            setValue('');
        }
    };

    const onFocusOut = (e) => {
        // restore default if no change is made
        if (value === '' && editable) {
            setValue(fallbackValue);
        }
        // set new fallback value
        if (value !== '') {
            setFallbackValue(e.target.value);
        }
    };
    if (!value || value === '') {
        return null;
    } else if (!editable) {
        return (
            <div className={classNames(styles.value, className)}>{value}</div>
        );
    }

    return maxInput ? (
        <TextareaAutosize
            readOnly={!editable}
            data-test-id={`${dataKey}-input`}
            className={classNames(styles.value, className, {
                [styles.active]: editable,
            })}
            value={value}
            placeholder={fallbackValue ? fallbackValue : placeholder}
            onBlur={onFocusOut}
            onChange={(e) => {
                if (e.target.value.length < 256) setValue(e.target.value);
            }}
            maxLength={256}
        />
    ) : (
        <input
            readOnly={!editable}
            data-test-id={`${dataKey}-input`}
            className={classNames(styles.value, className, {
                [styles.active]: editable,
            })}
            value={value}
            placeholder={fallbackValue ? fallbackValue : placeholder}
            onFocus={onFocusIn}
            onBlur={onFocusOut}
            onChange={(e) => setValue(e.target.value)}
        />
    );
};

SingleValue.propTypes = {
    value: PropTypes.string.isRequired,
    setValue: PropTypes.func,
    editable: PropTypes.bool,
    className: PropTypes.string,
    placeholder: PropTypes.string,
    dataKey: PropTypes.string,
};

export default SingleValue;
