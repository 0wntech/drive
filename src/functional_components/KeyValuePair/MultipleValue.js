import React from 'react';
import PropTypes from 'prop-types';

import X from '../../assets/svgIcons/X';
import SingleValue from './SingleValue';
import styles from './KeyValuePair.module.css';

const MultipleValue = ({
    dataKey,
    values,
    editable,
    setValue,
    placeholder,
}) => {
    return (
        <div className={styles.multiValueContainer}>
            {values.map((value, index) => (
                <div key={index} className={styles.valueWrapper}>
                    <SingleValue
                        dataKey={dataKey}
                        value={value}
                        editable={editable}
                        setValue={(value) => {
                            const valueCopy = [...values];
                            valueCopy[index] = value;
                            console.log('set value: ', dataKey, valueCopy);
                            setValue(dataKey, valueCopy);
                        }}
                        placeholder={placeholder}
                    />
                    {editable ? (
                        <X
                            className={styles.deleteIcon}
                            onClick={() => {
                                const valueCopy = [...values];
                                valueCopy.splice(index, 1);
                                setValue(dataKey, valueCopy);
                            }}
                        />
                    ) : null}
                </div>
            ))}
        </div>
    );
};

MultipleValue.propTypes = {
    dataKey: PropTypes.string,
    values: PropTypes.array,
    editable: PropTypes.bool,
    setValue: PropTypes.func,
    placeholder: PropTypes.string,
};

export default MultipleValue;
