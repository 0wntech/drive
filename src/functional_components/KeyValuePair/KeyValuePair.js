import React from 'react';
import PropTypes from 'prop-types';
import { Collapse } from 'react-collapse';

import styles from './KeyValuePair.module.scss';
import classNames from 'classnames';
import MultipleValue from './MultipleValue';
import SingleValue from './SingleValue';

const showAddButton = (value, editable) => {
    // multiple values possible (array) and at least one is already filled
    return typeof value === 'object' && editable;
};

const KeyValuePair = ({
    dataKey,
    editable,
    label,
    placeholder,
    setValue,
    value,
    maxInput,
}) => {
    console.log(value);
    const renderValues = (value) => {
        if (typeof value === 'object') {
            // is an array
            return (
                <MultipleValue
                    dataKey={dataKey}
                    values={value}
                    editable={editable}
                    setValue={setValue}
                    placeholder={placeholder}
                />
            );
        } else {
            // is a single value
            return (
                <SingleValue
                    setValue={(value) => setValue(dataKey, value)}
                    dataKey={dataKey}
                    value={value}
                    placeholder={placeholder}
                    editable={editable}
                    maxInput={maxInput}
                />
            );
        }
    };

    return value || editable ? (
        <div className={styles.container}>
            <div className={styles.keyValueContainer}>
                <div className={styles.keyLabel}>{label}</div>
                <div className={styles.valueContainer}>
                    {renderValues(value)}
                </div>
            </div>

            <Collapse isOpened={showAddButton(value, editable)}>
                <div
                    className={classNames(
                        styles.containerFooter,
                        styles.addLabel,
                        { [styles.active]: editable }
                    )}
                    onClick={() => setValue(dataKey, [...value, ''])}
                >
                    + Add
                </div>
            </Collapse>
        </div>
    ) : null;
};

KeyValuePair.propTypes = {
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
    dataKey: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    setValue: PropTypes.func,
    editable: PropTypes.bool,
};

export default KeyValuePair;
