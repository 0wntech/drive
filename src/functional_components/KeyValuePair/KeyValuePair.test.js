import React from 'react';
import KeyValuePair from './KeyValuePair';
import { shallow } from 'enzyme';
import TestRenderer from 'react-test-renderer';

describe('KeyValuePair', () => {
    const label = 'Test';
    const value = 'testValue';
    const dataKey = 'testKey';
    const placeholder = 'testPlaceholder';

    it('should render unexpanded', () => {
        const editable = false;
        const tree = TestRenderer.create(
            <KeyValuePair
                placeholder={placeholder}
                dataKey={dataKey}
                value={value}
                label={label}
                editable={editable}
            />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should render expanded', () => {
        const editable = true;
        const tree = TestRenderer.create(
            <KeyValuePair
                editable={editable}
                placeholder={placeholder}
                dataKey={dataKey}
                value={value}
                label={label}
            />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    xit('should trigger onClick function on click', () => {
        const mockFunction = jest.fn();
        const wrapper = shallow(
            <KeyValuePair label={label} onClick={mockFunction} />
        );
        wrapper.simulate('click');
        expect(mockFunction).toHaveBeenCalled();
    });
});
