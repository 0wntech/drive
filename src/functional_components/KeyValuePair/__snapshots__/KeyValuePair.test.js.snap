// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValuePair should render expanded 1`] = `
<div
  className="container"
>
  <div
    className="keyValueContainer"
  >
    <div
      className="keyLabel"
    >
      Test
    </div>
    <div
      className="valueContainer"
    >
      <input
        className="value active"
        data-test-id="testKey-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="testValue"
        readOnly={false}
        value="testValue"
      />
    </div>
  </div>
  <div
    className="ReactCollapse--collapse"
    style={
      Object {
        "height": "0px",
        "overflow": "hidden",
      }
    }
  >
    <div
      className="ReactCollapse--content"
    >
      <div
        className="containerFooter addLabel active"
        onClick={[Function]}
      >
        + Add
      </div>
    </div>
  </div>
</div>
`;

exports[`KeyValuePair should render uneditable with multiple values 1`] = `
<div
  className="container"
>
  <div
    className="keyValueContainer"
  >
    <div
      className="keyLabel"
    >
      Test
    </div>
    <div
      className="valueContainer"
    >
      <div
        className="multiValueContainer"
      >
        <div
          className="valueWrapper"
        >
          <input
            className="value"
            data-test-id="testKey-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="multiple"
            readOnly={true}
            value="multiple"
          />
        </div>
        <div
          className="valueWrapper"
        >
          <input
            className="value"
            data-test-id="testKey-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="values"
            readOnly={true}
            value="values"
          />
        </div>
        <div
          className="valueWrapper"
        >
          <input
            className="value"
            data-test-id="testKey-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="test"
            readOnly={true}
            value="test"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ReactCollapse--collapse"
    style={
      Object {
        "height": "0px",
        "overflow": "hidden",
      }
    }
  >
    <div
      className="ReactCollapse--content"
    >
      <div
        className="containerFooter addLabel"
        onClick={[Function]}
      >
        + Add
      </div>
    </div>
  </div>
</div>
`;

exports[`KeyValuePair should render unexpanded 1`] = `
<div
  className="container"
>
  <div
    className="keyValueContainer"
  >
    <div
      className="keyLabel"
    >
      Test
    </div>
    <div
      className="valueContainer"
    >
      <input
        className="value"
        data-test-id="testKey-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="testValue"
        readOnly={true}
        value="testValue"
      />
    </div>
  </div>
  <div
    className="ReactCollapse--collapse"
    style={
      Object {
        "height": "0px",
        "overflow": "hidden",
      }
    }
  >
    <div
      className="ReactCollapse--content"
    >
      <div
        className="containerFooter addLabel"
        onClick={[Function]}
      >
        + Add
      </div>
    </div>
  </div>
</div>
`;
